### DISCUSSION TITLE
.NET 8/ C# 12 changes

### DISCUSSION THREAD

**Discussion by TimothyMakkison** [riok/mapperly#675] [General]
As the release date of .NET 8 draws closer and finalized previews of C# 12 are released, I thought that it would be worth looking at ensuring Mapperly supports the new features. I've compiled some standout issues and potential additions that could be considered.

- Add .NET 8 preview/Roslyn 4.6
  - Add .NET 8/Roslyn 4.6 to the integration tests and upgrade the samples/unit tests to .NET 8 when it is released.
  - Add the integration test to the CI pipeline.
- Consider using the new unsafe accessor feature.
  - See #600 for accessing private members. I have a working version although it is not ready to be added.
  - It is also possible to access private constructors and bypass `init`/`required` members, this could be further explored (perhaps for setting nested `init`/`required` properties).
- Look at using `CollectionMarshal.SetCount` and `CollectionMarshal.AsSpan` for extremely fast list copying (technically breaking).
- Ensure that [type aliasing](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-12#alias-any-type) is supported, unit tests should be added.
- Add support for primary constructors, I'd assume that they have priority over non `MapperConstructor` marked constructors.

**Comment by latonz** [comment]
Created https://github.com/riok/mapperly/issues/682, https://github.com/riok/mapperly/issues/683 and https://github.com/riok/mapperly/issues/684. Overview for .NET 8.0 related work: https://github.com/riok/mapperly/issues/689

Improvements with `CollectionMarshal` may need some experiments...

**Reply by latonz** [reply]
Closing this for now as relevant issues are created. Integration tests are already running on .NET 8.0.
If the experiments with `CollectionMarshal` bring any actionable results or sth. to discuss feel free to open a new discussion.
